src/activity/activity_heartbeat_manager.rs

```diff
@@ -1,86 +1,119 @@
     shutdown_tx: Sender<bool>,
+    shutdown_rx: Receiver<bool>,
 }
 
-struct ActivityHeartbeatTask {
-    task_token: Vec<u8>,
-    next_delay: time::Duration,
-    trigger: Sleep,
+struct ActivityHeartbeatProcessorHandle {
+    heartbeat_tx: Sender<Vec<common::Payload>>,
+    join_handle: JoinHandle<()>,
+}
+
+struct ActivityHeartbeatProcessor {
+    delay: time::Duration,
+    heartbeat_rx: Receiver<Vec<common::Payload>>,
+    shutdown_rx: Receiver<bool>,
 }
 
 impl ActivityHeartbeatManager {
     pub fn new() -> Self {
-        let (heartbeat_tx, heartbeat_rx) = unbounded::<ActivityHeartbeatTask>();
-        let (shutdown_tx, shutdown_rx) = bounded(1);
-        thread::spawn(move || Self::processor(heartbeat_rx, shutdown_rx));
+        let (shutdown_tx, shutdown_rx) = channel(false);
         Self {
-            heartbeat_details: Default::default(),
-            scheduled_heartbeats: Default::default(),
-            heartbeat_tx,
+            heartbeat_processors: Default::default(),
             shutdown_tx,
+            shutdown_rx,
         }
     }
 
     pub fn record(&self, heartbeat: ActivityHeartbeat) -> Result<(), ActivityHeartbeatError> {
-        self.heartbeat_details
-            .insert(heartbeat.task_token.clone(), heartbeat.details);
-
-        if self
-            .scheduled_heartbeats
-            .insert(heartbeat.task_token.clone())
-        {
-            let initial_delay = time::Duration::from_millis(0);
-            let heartbeat_timeout: time::Duration = heartbeat
-                .heartbeat_timeout
-                .ok_or(ActivityHeartbeatError::HeartbeatTimeoutNotSet)?
-                .try_into()
-                .or(Err(ActivityHeartbeatError::InvalidHeartbeatTimeout))?;
-            let next_delay = heartbeat_timeout.div(2);
-
-            self.heartbeat_tx
-                .send(ActivityHeartbeatTask {
-                    task_token: heartbeat.task_token,
-                    next_delay,
-                    trigger: sleep(initial_delay),
-                })
-                .expect("failed to send activity task into the channel");
-        };
-
+        /// Do not accept new heartbeat requests if core is shutting down, this is because we
+        /// don't want to spawn new processors and existing processors won't be accepting new work.
+        if *self.shutdown_rx.borrow() {
+            return Err(ActivityHeartbeatError::ShuttingDown);
+        }
+        match self.heartbeat_processors.get(&heartbeat.task_token) {
+            Some(handle) => {
+                handle.heartbeat_tx.send(heartbeat.details);
+            }
+            None => {
+                let heartbeat_timeout: time::Duration = heartbeat
+                    .heartbeat_timeout
+                    .ok_or(ActivityHeartbeatError::HeartbeatTimeoutNotSet)?
+                    .try_into()
+                    .or(Err(ActivityHeartbeatError::InvalidHeartbeatTimeout))?;
+                let delay = heartbeat_timeout.div(2);
```

Comment:
This delay factor should probably be configurable

Reply:
Oh, I see why it isn't at the moment -- it might be nice to do the two different timeouts thing I mentioned which is still pretty simple, where there's one delay for "minimum interval between beats" and another delay for "give up if you haven't seen a heartbeat in this long". Not a huge deal tho

Reply:
I suggest we keep it simple for now, as both are the same and separate them if we ever need to set different values.